includes/class-amp-theme-support.php

```diff
@@ -1934,17 +1934,28 @@ public static function prepare_response( $response, $args = [] ) {
 			);
 		}
 
-		// Abort if an expected template was not rendered, in that template actions didn't fire and response type is not HTML.
-		$did_template_action = (
+		$amp_attribute_pattern = sprintf(
+			'#^(?:<!.*?>|\s+)*+<html[^>]*?\s(%1$s|%2$s|%3$s)(?:=".*?")?.*?(?>>.*?<)head[^>]*?>#is',
+			Attribute::AMP,
+			Attribute::AMP_EMOJI,
+			Attribute::AMP_EMOJI_ALT
+		);
+
+		$should_be_processed = (
 			did_action( 'wp_head' )
 			||
 			did_action( 'wp_footer' )
 			||
 			did_action( 'amp_post_template_head' )
 			||
 			did_action( 'amp_post_template_footer' )
+			||
+			preg_match( $amp_attribute_pattern, $response )
 		);
-		if ( ! $did_template_action || Attribute::TYPE_HTML !== substr( AMP_HTTP::get_response_content_type(), 0, 9 ) ) {
+
+		// Abort if an expected template action didn't fire, the HTML tag does not have the AMP attribute or if the
+		// response type is not HTML.
+		if ( ! $should_be_processed || Attribute::TYPE_HTML !== substr( AMP_HTTP::get_response_content_type(), 0, 9 ) ) {
 			return $response;
 		}
```

Comment:
This is a minor performance tweak, but we could check the HTML content type in a separate conditional above, which would avoid the need to do any other conditionals later. For example:

```php
// Abort if response type is not HTML.
if ( Attribute::TYPE_HTML !== substr( AMP_HTTP::get_response_content_type(), 0, 9 ) ) {
	return $response;
}

// Abort if an expected template action didn't fire or if the HTML tag does not have the AMP attribute.
if ( ! (
	did_action( 'wp_head' )
	||
	did_action( 'wp_footer' )
	||
	did_action( 'amp_post_template_head' )
	||
	did_action( 'amp_post_template_footer' )
	||
	preg_match(
		sprintf(
			'#^(?:<!.*?>|\s+)*+<html[^>]*?\s(%1$s|%2$s|%3$s)(?:=".*?")?.*?(?>>.*?<)head[^>]*?>#is',
			Attribute::AMP,
			Attribute::AMP_EMOJI,
			Attribute::AMP_EMOJI_ALT
		),
		$response
	)
) ) {
	return $response;
}
```

Reply:
Done in https://github.com/ampproject/amp-wp/pull/5316/commits/acff806548129bc6e0933e2b4f114b6fcf8728e6.